File: set_ns.py
import os
import sys

import yaml

CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
ANNOTATOR_PATH = os.path.join(CURRENT_DIR, "config", "overlays", "namespaced", "roles-ns-annotator.yaml")
MANAGER_PATCH_PATH = os.path.join(CURRENT_DIR, "config", "overlays", "namespaced", "namespaced-manager-patch.yaml")


def update_annotator_file(namespaces_to_set):
    # Load the YAML file
    with open(ANNOTATOR_PATH, 'r') as file:
        doc = yaml.safe_load(file)

    doc['values'] = [ns.strip() for ns in namespaces_to_set.split(',')]

    with open(ANNOTATOR_PATH, 'w') as file:
        yaml.dump(doc, file, default_flow_style=False)

def set_watch_namespace_env_var(namespaces_to_set):
    with open(MANAGER_PATCH_PATH, 'r') as file:
        doc = yaml.safe_load(file)
    
    # Dicts are always shallow copies and our patch is limited to one
    # one container and one environement value
    containers_spec = doc["spec"]["template"]["spec"]["containers"]
    containers_spec[0]["env"][0]["value"] = namespaces_to_set

    with open(MANAGER_PATCH_PATH, 'w') as file:
        yaml.dump(doc, file, default_flow_style=False)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: {} \"namespace1,namespace2\"")
        sys.exit(1)

    namespaces_to_set = sys.argv[1]
    
    update_annotator_file(namespaces_to_set)
    set_watch_namespace_env_var(namespaces_to_set)

File: set_ns/set_ns.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

var currentDir string

func init() {
	var err error
	currentDir, err = os.Getwd()
	if err != nil {
		panic(fmt.Sprintf("Error getting current directory: %v", err))
	}
}

var (
	annotatorPath    = filepath.Join(currentDir, "..", "config", "overlays", "namespaced", "roles-ns-annotator.yaml")
	managerPatchPath = filepath.Join(currentDir, "..", "config", "overlays", "namespaced", "namespaced-manager-patch.yaml")
)

func updateAnnotatorFile(namespacesToSet string) error {
	data, err := os.ReadFile(annotatorPath)
	if err != nil {
		return err
	}

	var doc map[string]interface{}
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return err
	}

	namespaces := strings.Split(namespacesToSet, ",")
	for i := range namespaces {
		namespaces[i] = strings.TrimSpace(namespaces[i])
	}
	doc["values"] = namespaces

	out, err := yaml.Marshal(doc)
	if err != nil {
		return err
	}

	return os.WriteFile(annotatorPath, out, 0644)
}

func setWatchNamespaceEnvVar(namespacesToSet string) error {
	data, err := os.ReadFile(managerPatchPath)
	if err != nil {
		return err
	}

	var doc map[string]interface{}
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return err
	}

	containers := doc["spec"].(map[string]interface{})["template"].(map[string]interface{})["spec"].(map[string]interface{})["containers"].([]interface{})
	container := containers[0].(map[string]interface{})
	envList := container["env"].([]interface{})
	envVar := envList[0].(map[string]interface{})

	envVar["value"] = namespacesToSet

	out, err := yaml.Marshal(doc)
	if err != nil {
		return err
	}

	return os.WriteFile(managerPatchPath, out, 0644)
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: ./app \"namespace1,namespace2\"")
		os.Exit(1)
	}

	namespacesToSet := os.Args[1]

	if err := updateAnnotatorFile(namespacesToSet); err != nil {
		fmt.Printf("Error updating annotator file: %v\n", err)
		os.Exit(1)
	}

	if err := setWatchNamespaceEnvVar(namespacesToSet); err != nil {
		fmt.Printf("Error updating manager patch file: %v\n", err)
		os.Exit(1)
	}
}
